language/resource_types.md

```diff
@@ -0,0 +1,502 @@
+
+### ensureable
+
+Creates an `ensure` property with acceptable values of `present` and `absent`, each of which invoke the provider's `create` and `destroy` methods, respectively:
+
+```ruby
+Puppet::Type.newtype(:ec2_instance) do
+  ensurable
+end
+```
+
+Ensurable can also take a block, and in that case, the type should define the allowed set of values. For example, the `package` type defines allowed values as:
+
+```ruby
+Puppet::Type.newtype(:package) do
+  ensurable do
+    newvalue(:present)
+    newvalue(:absent)
+    newvalue(:purged)
+    newvalue(:held)
+    newvalue(:latest)
+    newvalue(/./)
+  end
+end
+```
+
+### newproperty
+
+Defines a new property for a type. Puppet will ensure the resource's current state (as retrieved by the provider) matches the property's desired state (as expressed in the manifest).
+
+```ruby
+Puppet::Type.newtype(:ec2_instance) do
+  newproperty(:region) do
+  end
+end
+```
+
+### newparam
+
+Defines a new parameter for the type. Puppet does not enforce state for a parameter. Instead parameters specify additional information about how the provider should ensure the resource and its properties are in the correct state.
+
+```ruby
+Puppet::Type.newtype(:ec2_instance) do
+  newparam(:user_data) do
+  end
+end
+```
+
+### newmetaparam
+
+Defines a new metaparameter for a type. Puppet has several built-in metaparameters described below. It is uncommon for types to define new metaparams as these generally require puppet core changes to be useful, e.g. `noop`, `tag`, `schedule`, etc.
+
+### validate
+
+Performs per-resource validation at catalog application time. This method is often used to validate related parameters and properties for a single resource, e.g. mutually exclusive properties or if one property requires another. One concrete example is the `file` resource requires the `source` parameter to be set if the `recurse` parameter is set to `remote`:
+
+```ruby
+Puppet::Type.newtype(:file) do
+  validate do
+    self.fail "You cannot specify a remote recursion without a source" if !self[:source] && self[:recurse] == :remote
+  end
+end
+```
+
+### autorelations
+
+Allows a type to define automatic relationships (`before`, `subscribe`, `require`, `notify`) between each instance of the type and instances of a different type. For example, the `exec` type will autorequire any file resource whose `path` matches the exec  `command`. Note the relationship will only be added to the catalog if puppet is managing both ends of the relationship. That way puppet will automatically create the file containing the command to execute, before trying to executing it, regardless of manifest order.
+
```

Comment:
If you're asking about why autorelations are conditionally created when the autorelated resource is in the catalog, this is because it allows opportunistic relationships to be established. For instance, mounts automatically try to depend on any mounts that are higher up in the file system (https://github.com/puppetlabs/puppet/blob/4.4.1/lib/puppet/type/mount.rb#L287-L294) but don't interrogate the catalog to make sure the other mounts are there. This offloads the complexity from the definition of the autorelations to the actual generation of edges.

The same thing goes for execs - we can opportunistically depend on files that the exec depends on without enforcing that the file is managed, nor do we have to manually interrogate the catalog to make sure the file is there before adding the dependency.

It may be up for debate as to whether this is a good idea but that's probably a lengthy topic in and of itself. If someone has the enthusiasm to discuss this on the dev list and then add support for mandatory autorelationships that would be ace, but for the sake of defining the current behavior I think we're best off focusing on getting the current behavior and rationales documented.